src/shared/Paramable.ts

```diff
@@ -154,45 +164,60 @@ export class Paramable implements ParamableInterface {
      * checkParameters() instead.
      */
     validate(): ValidationStatus {
-        const status = this.checkParameters()
-        this.isValid = status.isValid
-        if (this.isValid) {
-            this.assignParameters()
+        const realParams: {[key: string]: unknown} = {}
+        for (const prop in this.params) {
+            if (!Object.prototype.hasOwnProperty.call(this.params, prop))
+                continue
+            const p = this.params[prop]
+            if (typeof p.value !== 'string') p.value = `${p.value}`
+
+            realParams[prop] = typeFunctions[p.type].realize(
+                p.value as string
+            )
         }
-        return status
+        return this.checkParameters(realParams)
     }
     /**
-     * checkParameters should check that all parameters are well-formed,
-     * in-range, etc.
-     * @returns {ValidationStatus}
+     * Performs the same purpose as `validate()`, but takes as a parameter
+     * a list of realized values of the parameters, so that the function doesn't
+     * have to do this itself. This function should always be the one overloaded
+     * by classes which extend `Paramable`, and not `validate()`
```

Comment:
I can't make `validate` as private since it is still called externally. It is an exposed part of the Paramable interface, and the generic implementation `Paramable` implements the validate function to do some bookkeeping and then giving the main validation to the `checkParameters` function

Reply:
Ah, you want to make it "final" in Java terminology. See [this playground](https://www.typescriptlang.org/play/?target=99&jsx=0#code/PTAEBMEsGcEMCMA2BTUBjAFpR4BOyA7dDWAgc0nNFlAFtkAXDAe3AChxk1FZ91mC0BqAD6ALlABXApACOk1NACeteM0QBuNgyUAHVADFmueJHCciAXlABvUAG0RAXQk79zAGaiNoAL7a9VAA5ZgB5ADdkXFwzZAAeABVLcOYzAD5QawTQADJQIxMzCx8QUAARAQByYWQAD11jYSYYUDdkAEI2D2k0BkgBOlgAa2QQiKiYzkS0gApw2EQFCQSASgkxyOjY6dtQUuNWrGhQboJe-oJO0FB8BklcInnF1FhjjYnthLStX1A2Nm4r2OAAVeIRhDY2NdrvQmKwZmtQO8tlMCJJVFEMpDoTj+IJ1MgAHSIZhkGaVADqzAA7u1Kisobjrrd7kRaMNRmFNpNkDMAOwMnH+fxsYAAKlAwNuhHAh1QHnUJOplDIoBalGoBGYTCiJ2wqDFwABPGgxwAwlgcKA6gwZSCwQQIYzrqUAKpnZi0eiOlVy0DwZBKuiMFiy5WIRA3ZDQSSIYQamhoT26fXW6LGZ17CWZ2GhhESNEY3C2TM4lkPUAx-S4Qm5+ErUAAalAAGZM79DWwRUnBMI0JlQARkNTQBbsOAEQCBNACcTSeSpaxJGhkOBKgAadC1kP1hlAA) for an example of the apparently best extant typescript technique for making a class method non-overridable, from the final comment in https://github.com/microsoft/TypeScript/issues/50532. So if the design is that this method _shall not_ be overridden, please use this technique. If there is a case for derived classes to override it, please just include a strongly worded warning against overriding it and a brief hint as to the unusual circumstances in which one might override it. Thanks!